core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -350,7 +328,261 @@ public void run() {
+      public void run() {
+        // Fill out timeout on the headers
+        headers.removeAll(TIMEOUT_KEY);
+        // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
+        // because timeout tolerates the clock difference between machines.
+        Long timeoutMicros = getRemainingTimeoutMicros();
+        if (timeoutMicros != null) {
+          if (timeoutMicros <= 0) {
+            maybeClosePrematurely(Status.DEADLINE_EXCEEDED);
+            return;
+          }
+          headers.put(TIMEOUT_KEY, timeoutMicros);
+        }
+
+        if (transportFuture.isDone()) {
+          ClientTransport transport;
+          try {
+            transport = transportFuture.get();
+          } catch (Exception e) {
+            maybeClosePrematurely(Status.fromThrowable(e));
+            return;
+          }
+          createStream(transport);
+        } else {
+          transportFuture.addListener(streamCreationTask, callExecutor);
+        }
+      }
+
+      @Override
+      public String toString() {
+        return "streamCreationTask";
+      }
+    };
+
+    DelayedStream(Metadata headers, Listener<RespT> callListener) {
+      this.headers = headers;
+      listener = new ClientStreamListenerImpl(callListener);
+      transportFuture = clientTransportProvider.get();
+      if (transportFuture == null) {
+        maybeClosePrematurely(Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+        return;
+      }
+      streamCreationTask.run();
+    }
+
+    /**
+     * Creates a stream on a presumably usable transport.
+     */
+    private void createStream(ClientTransport transport) {
+      synchronized (this) {
+        if (realStream == NOOP_CLIENT_STREAM) {
+          // Already cancelled
+          return;
+        }
+        Preconditions.checkState(realStream == null, "Stream already created: %s", realStream);
+        try {
+          realStream = transport.newStream(method, headers, listener);
+        } catch (IllegalStateException e) {
+          // The transport is already shut down. This may due to a race condition between Channel
+          // shutting down idle transports. Retry streamCreationTask, but Schedule it in an executor
+          // to avoid recursion.
+          callExecutor.execute(streamCreationTask);
+          return;
+        }
+        if (compressor != null) {
+          realStream.setCompressor(compressor);
+        }
+        if (decompressor != null) {
+          if (decompressor instanceof Decompressor) {
+            realStream.setDecompressor((Decompressor) decompressor);
+          } else if (decompressor instanceof String) {
+            realStream.setDecompressor((String) decompressor);
+          } else {
+            throw new IllegalStateException("Unexpected decompressor " + decompressor);
+          }
+        }
+        if (this.decompressionRegistry != null) {
+          realStream.setDecompressionRegistry(this.decompressionRegistry);
```

Comment:
I like to keep `this`, so that if I misspelt this variable as `decompressorRegistry`, it wouldn't access that field from the outer class without my notice.